### PR TITLE
Arrays for Python compiler

### PR DESCRIPTION
**Pull request by prvshah51**
Arrays can be compiled to python now, Multilevel Arrays can be created too.

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [MIT license](https://github.com/dafny-lang/dafny/blob/master/LICENSE.txt).</small>

### CODE REVIEW

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 1]
```diff
@@ -500,8 +502,19 @@ protected class ClassWriter : IClassWriter {
...
+      return wBody;
+    }
+
     protected override ConcreteSyntaxTree CreateForLoop(string indexVar, string bound, ConcreteSyntaxTree wr) {
-      throw new NotImplementedException();
+      string headerFormat = "for {0} in range({1}):";
+      object[] headerArgs = { indexVar, bound };
+      return wr.NewBlockPy(string.Format(headerFormat, headerArgs), null);
+
+
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:517]:
```suggestion
      return wr.NewBlockPy($"for {indexVar} in range({bound}):");
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 2]
```diff
@@ -453,16 +457,14 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override ConcreteSyntaxTree CreateLabeledCode(string label, bool createContinueLabel, ConcreteSyntaxTree wr) {
-      if (createContinueLabel) { throw new NotImplementedException(); }
+      Contract.Assert(!createContinueLabel);
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:460]: This was suggested in review. Let's keep it that way.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 3]
```diff
@@ -453,16 +457,14 @@ protected class ClassWriter : IClassWriter {
...
       return wr.NewBlockPy($"with _dafny.label(\"{label}\"):");
     }
 
     protected override void EmitBreak(string label, ConcreteSyntaxTree wr) {
       if (label != null) {
         wr.WriteLine($"_dafny._break(\"{label}\")");
-      } else {
-        wr.WriteLine("break");
-      }
+      } else { wr.WriteLine("break"); }
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:467]: This was suggested in review. Let's keep it that way.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 4]
```diff
@@ -1,6 +1,7 @@
 using System;
 using System.Collections.Generic;
 using System.Collections.ObjectModel;
+using System.ComponentModel.DataAnnotations;
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:4]:
```suggestion
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 5]
```diff
@@ -323,6 +324,9 @@ protected class ClassWriter : IClassWriter {
             return "int";
           }
         case UserDefinedType udt: {
+            if (xType.IsArrayType) {
+              return "list";
+            }
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:329]: Let's fix this in `FullTypeName`.

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    // Conversion to sequence should copy elements (sequences are immutable!)
+    a[0] := 42;
+    print s, "\n";
+    MultipleDimensions();
+    
+    
+   }
+   
+   method MultipleDimensions() {
+     var matrix := new int[2,8];
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:61]: This doesn't create any output.

Comment (prvshah51) [reply Test/comp/firstSteps/7_Arrays.dfy:61]: I will add to that, it's just that the file has a user defined printMatrix function which is not working properly yet but this does make a proper array so I will add a print statement to it.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 2 above)
Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs:459]: so just to be sure are you suggesting to keep the line 456 as first line of the function, right ?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 7]
```diff
@@ -682,12 +718,30 @@ protected class ClassWriter : IClassWriter {
...
 
     protected override ConcreteSyntaxTree EmitArraySelect(List<Expression> indices, Type elmtType, bool inLetExprBody,
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      Contract.Assert(indices != null && 1 <= indices.Count);  // follows from precondition
+      var w = wr.Fork();
+      if (indices.Count == 1) {
+        wr.Write("[");
+        TrExpr(indices[0], wr, inLetExprBody, wStmts);
+        wr.Write("]");
```
[the remark this answers is not in the document]
Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs:734]: Good catch, at first I was using different code and forget to merger it all once I changed it back.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 8]
```diff
@@ -544,7 +557,21 @@ protected class ClassWriter : IClassWriter {
...
-      throw new NotImplementedException();
+      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : null;
+      if (dimensions.Count == 1) {
+        // handle the common case of 1-dimensional arrays separately
+        var arrayType = TypeInitializationValue(elmtType, wr, tok, false, false);
+        wr.Write($"[{arrayType} for _ in range");
+        TrParenExpr(dimensions[0], wr, false, wStmts);
+        wr.Write("]");
+      } else {
+        wr.Write("_dafny.newArray({0}", initValue ?? "None");
```
[the remark this answers is not in the document]
Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs:567]: oh Yes, Runtime file will be added in next commit.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 2 above)
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:459]: I think so, just don't change anything.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 9]
```diff
@@ -59,6 +60,7 @@ public class PythonCompiler : SinglePassCompiler {
     public override void EmitCallToMain(Method mainMethod, string baseName, ConcreteSyntaxTree wr) {
       Coverage.EmitSetup(wr);
       wr.NewBlockPy("try:")
+
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:63]:
```suggestion
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 10]
```diff
@@ -800,9 +819,10 @@ protected class ClassWriter : IClassWriter {
         //TODO: Add deeper types
         return "Callable";
       }
-
       var cl = udt.ResolvedClass;
+
       return cl switch {
+        ArrayClassDecl => "list",
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:825]: Going by your Runtime, this should be an Array. Is there a reason not to make this a sequence?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 11]
```diff
@@ -686,7 +688,21 @@ protected class ClassWriter : IClassWriter {
 
     protected override void EmitNewArray(Type elmtType, IToken tok, List<Expression> dimensions, bool mustInitialize,
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : null;
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:691]:
```suggestion
      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : "None";
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 12]
```diff
@@ -686,7 +688,21 @@ protected class ClassWriter : IClassWriter {
...
+      if (dimensions.Count == 1) {
+        // handle the common case of 1-dimensional arrays separately
+        var arrayType = TypeInitializationValue(elmtType, wr, tok, false, false);
+        wr.Write($"[{initValue} for _ in range");
+        TrParenExpr(dimensions[0], wr, false, wStmts);
+        wr.Write("]");
+      } else {
+        wr.Write("{0}.newArray({1}", DafnyArrayClass, initValue ?? "None");
+        foreach (var dim in dimensions) {
+          wr.Write(",int ");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:701]:
```suggestion
          wr.Write(", int");
```
But I think we should just skip those.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 13]
```diff
@@ -711,6 +727,9 @@ protected class ClassWriter : IClassWriter {
             case BigDec n:
               wr.Write($"{DafnyRuntimeModule}.BigRational('{n.Mantissa}e{n.Exponent}')");
               break;
+            case null:
+              wr.Write("");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:731]:
```suggestion
              wr.Write("None");
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 14]
```diff
@@ -856,6 +876,18 @@ protected class ClassWriter : IClassWriter {
...
+        case SpecialField.ID.ArrayLengthInt:
+          preString = "len(";
+          if (idParam == null) {
+            postString = ")";
+          } else {
+            if ((int)idParam == 0) {
+              postString = ")";
+            }
+            postString = String.Concat(Enumerable.Repeat("[0]", (int)idParam)) + ")";
+          }
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:889]: This isn't needed to pass the tests.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 15]
```diff
@@ -1098,6 +1150,11 @@ protected class ClassWriter : IClassWriter {
           staticCallString = $"{DafnyRuntimeModule}.euclidian_modulus"; break;
 
         case BinaryExpr.ResolvedOpcode.EqCommon:
+          opString = "=="; break;
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:1153]: What's the point?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 16]
```diff
@@ -908,12 +940,24 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override ConcreteSyntaxTree EmitArraySelect(List<string> indices, Type elmtType, ConcreteSyntaxTree wr) {
-      throw new NotImplementedException();
+      Contract.Assert(indices != null && 1 <= indices.Count);
+      var w = wr.Fork();
+      if (indices.Count == 1) {
+        wr.Write("[{0}]", indices[0]);
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:946]:
```suggestion
        wr.Write($"[{indices[0]}]");
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 17]
```diff
@@ -908,12 +940,24 @@ protected class ClassWriter : IClassWriter {
...
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      Contract.Assert(indices != null && 1 <= indices.Count);  // follows from precondition
+      var w = wr.Fork();
+      foreach (var index in indices) {
+        wr.Write("[");
+        TrExpr(index, wr, inLetExprBody, wStmts);
+        wr.Write("]");
+      }
+      return w;
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:960]:
```suggestion
      var strings = indices.Select(index => Expr(indices[0], inLetExprBody, wStmts).ToString());
      return EmitArraySelect(strings.ToList(), elmtType, wr);
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 18]
```diff
@@ -936,7 +980,15 @@ protected class ClassWriter : IClassWriter {
...
-      throw new NotImplementedException();
+      wr.Write($"{DafnySeqClass}(");
+      TrParenExpr(source, wr, inLetExprBody, wStmts);
+      wr.Write("[");
+      if (lo != null) { TrExpr(lo, wr, inLetExprBody, wStmts); }
+      wr.Write(":");
+      if (hi != null) { TrExpr(hi, wr, inLetExprBody, wStmts); }
+      wr.Write(":");
+      wr.Write("]");
+      wr.Write(")");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:991]:
```suggestion
      wr.Write(":])");
```

**Review by fabiomadge — COMMENTED** [review]
Please clean up some of the white space.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 14 above)
Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs:889]: It is used by the file `7_Arrays.dfy` test file.

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+   }
+   
+   method PrintArrayArray<A>(a: array<array<A>>) {
+     var i := 0;
+     while i < a.Length {
+       print a[i][..], " ";
+       i := i + 1;
+     }
+     print "\n";
+   }
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:93]:
```suggestion
   }

```

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    s := a[20..];
+    print s, "\n";
+    s := a[..8];
+    print s, "\n";
+    
+    // Conversion to sequence should copy elements (sequences are immutable!)
+    a[0] := 42;
+    print s, "\n";
+    MultipleDimensions();
+    }
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:58]:
```suggestion
  }
```

File: Source/DafnyRuntime/DafnyRuntime.py [hunk 21]
```diff
@@ -155,6 +155,16 @@ def euclidian_modulus(a, b):
...
     return c if c == 0 else bp - c
 
+def newArray(initValue, *dims):
+    b = initValue
+    dimslist = list(dims)
+    for i in reversed(dimslist):
+        tmplist = []
+        for _ in range(i):
+            tmplist.append(b)
+        b= tmplist
```
Comment (fabiomadge) [inline Source/DafnyRuntime/DafnyRuntime.py:165]:
```suggestion
        b = tmplist
```

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 17 above)
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:957]: ?

File: Test/comp/firstSteps/7_Arrays.dfy.expect [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 
+17
+[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22]
+[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+[20, 21, 22]
+[0, 1, 2, 3, 4, 5, 6, 7]
+[0, 1, 2, 3, 4, 5, 6, 7]
+0 0 0 0 0 0 0 0 
+0 0 0 0 0 0 0 0 
+[] [0] [0, 0] [0, 0, 0] [0, 0, 0, 0]
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy.expect:14]:
```suggestion
[] [0] [0, 0] [0, 0, 0] [0, 0, 0, 0] 
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 23]
```diff
@@ -856,6 +873,18 @@ protected class ClassWriter : IClassWriter {
...
+        case SpecialField.ID.ArrayLengthInt:
+          preString = "len(";
+          if (idParam == null) {
+            postString = ")";
+          } else {
+            if ((int)idParam == 0) {
+              postString = ")";
+            }
+            postString = String.Concat(Enumerable.Repeat("[0]", (int)idParam)) + ")";
+          }
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:886]:
```suggestion
          postString = ")";
          if (idParam != null && (int)idParam > 0) {
            // TODO: Store Dimensions
            postString = string.Concat(Enumerable.Repeat("[0]", (int)idParam)) + postString;
          }
```

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 10 above)
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:822]: ?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 24]
```diff
@@ -800,9 +816,10 @@ protected class ClassWriter : IClassWriter {
         //TODO: Add deeper types
         return "Callable";
       }
-
       var cl = udt.ResolvedClass;
+
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:820]:
```suggestion
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 25]
```diff
@@ -800,9 +816,10 @@ protected class ClassWriter : IClassWriter {
         //TODO: Add deeper types
         return "Callable";
       }
-
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:803]: Can you add that line again?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 26]
```diff
@@ -686,7 +685,21 @@ protected class ClassWriter : IClassWriter {
 
     protected override void EmitNewArray(Type elmtType, IToken tok, List<Expression> dimensions, bool mustInitialize,
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : "None";
+      if (dimensions.Count == 1) {
+        // handle the common case of 1-dimensional arrays separately
+        var arrayType = TypeInitializationValue(elmtType, wr, tok, false, false);
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:691]:
```suggestion
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 27]
```diff
@@ -686,7 +685,21 @@ protected class ClassWriter : IClassWriter {
...
+        wr.Write($"[{initValue} for _ in range");
+        TrParenExpr(dimensions[0], wr, false, wStmts);
+        wr.Write("]");
+      } else {
+        wr.Write("_dafny.newArray({0}", initValue ?? "None");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:696]:
```suggestion
        wr.Write("_dafny.newArray({0}", initValue);
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 28]
```diff
@@ -686,7 +685,21 @@ protected class ClassWriter : IClassWriter {
...
+        foreach (var dim in dimensions) {
+          wr.Write(",int");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:698]: I don't think that we need those `int`s, but if you insist on them, please get the spacing right.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 29]
```diff
@@ -150,7 +150,6 @@ public class PythonCompiler : SinglePassCompiler {
     protected override IClassWriter CreateClass(string moduleName, string name, bool isExtern, string fullPrintName,
         List<TypeParameter> typeParameters, TopLevelDecl cls, List<Type> superClasses, IToken tok, ConcreteSyntaxTree wr) {
       var methodWriter = wr.NewBlockPy(header: $"class {IdProtect(name)}:");
-
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:153]: Same for this line.

File: Test/comp/firstSteps/7_Arrays.dfy.expect (hunk 22 above)
Comment (prvshah51) [reply Test/comp/firstSteps/7_Arrays.dfy.expect:14]: Can I know , why are you suggesting this ? This is the `.expect` file which is passing the test.

Comment (fabiomadge) [reply Test/comp/firstSteps/7_Arrays.dfy.expect:14]: Yes, have a look at how the test file for an explanation. The xUnit tests should have caught this, which I'll look into.

Comment (fabiomadge) [reply Test/comp/firstSteps/7_Arrays.dfy.expect:14]: https://github.com/dafny-lang/dafny/pull/2090

Comment (prvshah51) [reply Test/comp/firstSteps/7_Arrays.dfy.expect:14]: Thanks, It was actually not correct, IDE removed extra space during formatting and that code was committed.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 15 above)
Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs]: For all the non collection types `EqCommon` is used.

Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs]: Sure, but you didn’t group it in when I made that comment.

Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs]: Yes, it didn't cross my mind to check for it first.

File: Source/Dafny/Compilers/Compiler-python.cs (hunk 17 above)
Comment (fabiomadge) [reply Source/Dafny/Compilers/Compiler-python.cs:957]: It already works on my branch. I’d rather not have this duplication.

Comment (prvshah51) [reply Source/Dafny/Compilers/Compiler-python.cs:957]: Does the test in my PR works with this code? or might need to add If condition.

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 30]
```diff
@@ -908,12 +937,24 @@ protected class ClassWriter : IClassWriter {
     }
 
     protected override ConcreteSyntaxTree EmitArraySelect(List<string> indices, Type elmtType, ConcreteSyntaxTree wr) {
-      throw new NotImplementedException();
+      Contract.Assert(indices != null && 1 <= indices.Count);
+      var w = wr.Fork();
+      if (indices.Count == 1) {
+        wr.Write($"[{indices[0]}]");
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:943]:
```suggestion
      foreach (var index in indices) {
        wr.Write($"[{index}]");
```

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 31]
```diff
@@ -686,7 +686,20 @@ protected class ClassWriter : IClassWriter {
...
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : "None";
+      if (dimensions.Count == 1) {
+        // handle the common case of 1-dimensional arrays separately
+        wr.Write($"[{initValue} for _ in range");
+        TrParenExpr(dimensions[0], wr, false, wStmts);
+        wr.Write("]");
+      } else {
+        wr.Write("_dafny.newArray({0}", initValue);
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:696]:
```suggestion
        wr.Write($"{DafnyRuntimeModule}.newArray({initValue}");
```
Ideally you would see those things yourself.

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 32]
```diff
@@ -0,0 +1,93 @@
...
+      i := i + 1;
+    }
+    print "\n";
+  }
+}
+
+method Main() {
+  var a := new int[23];
+  var i := 0;
+    while i < 23 {
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:38]: This indentation is strange.

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 33]
```diff
@@ -0,0 +1,93 @@
...
+   
+method PrintArrayArray<A>(a: array<array<A>>) {
+  var i := 0;
+  while i < a.Length {
+    print a[i][..], " ";
+    i := i + 1;
+  }
+  print "\n";
+}
+
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:93]:
```suggestion

```

File: Source/DafnyRuntime/DafnyRuntime.py [hunk 34]
```diff
@@ -155,6 +155,16 @@ def euclidian_modulus(a, b):
...
     return c if c == 0 else bp - c
 
+def newArray(initValue, *dims):
+    b = initValue
+    dimslist = list(dims)
+    for i in reversed(dimslist):
+        tmplist = []
+        for _ in range(i):
+            tmplist.append(b)
+        b = tmplist
```
Comment (fabiomadge) [inline Source/DafnyRuntime/DafnyRuntime.py:165]:
```suggestion
    for i in reversed(list(dims)):
        b = [copy.deepcopy(b) for _ in range(i)]
```
This doesn't do what we though it did.

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 35]
```diff
@@ -0,0 +1,92 @@
...
+}
+   
+method PrintArrayArray<A>(a: array<array<A>>) {
+  var i := 0;
+  while i < a.Length {
+    print a[i][..], " ";
+    i := i + 1;
+  }
+  print "\n";
+}
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:92]: Can we add a newline here, so we finally get rid of this red thing?

File: Test/comp/firstSteps/7_Arrays.dfy [hunk 36]
```diff
@@ -0,0 +1,92 @@
...
+
+method Main() {
+  var a := new int[23];
+  var i := 0;
+  while i < 23 {
+    a[i] := i;
+    i := i + 1;
+  }
+    PrintArray(a);
```
Comment (fabiomadge) [inline Test/comp/firstSteps/7_Arrays.dfy:42]: ?

File: Source/Dafny/Compilers/Compiler-python.cs [hunk 37]
```diff
@@ -686,7 +686,20 @@ protected class ClassWriter : IClassWriter {
...
       ConcreteSyntaxTree wr, ConcreteSyntaxTree wStmts) {
-      throw new NotImplementedException();
+      var initValue = mustInitialize ? DefaultValue(elmtType, wr, tok, true) : "None";
+      if (dimensions.Count == 1) {
+        // handle the common case of 1-dimensional arrays separately
+        wr.Write($"[{initValue} for _ in range");
+        TrParenExpr(dimensions[0], wr, false, wStmts);
+        wr.Write("]");
+      } else {
+        wr.Write($"{DafnyRuntimeModule}.newArray({0}", initValue);
```
Comment (fabiomadge) [inline Source/Dafny/Compilers/Compiler-python.cs:696]: I‘m pretty sure I asked you to in-line this in the previous one as well.

**Review by fabiomadge — APPROVED** [review]

**Pull request merged**